src/lib/storage/dictionary_segment.hpp

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#include <limits>
+#include <memory>
+#include <iostream>
+#include <string>
+#include <utility>
+#include <vector>
+#include <iterator>
+#include <set>
+
+#include "all_type_variant.hpp"
+#include "type_cast.hpp"
+#include "types.hpp"
+#include "base_attribute_vector.hpp"
+#include "base_segment.hpp"
+#include "fixed_size_attribute_vector.hpp"
+
+namespace opossum {
+
+// Even though ValueIDs do not have to use the full width of ValueID (uint32_t), this will also work for smaller ValueID
+// types (uint8_t, uint16_t) since after a down-cast INVALID_VALUE_ID will look like their numeric_limit::max()
+constexpr ValueID INVALID_VALUE_ID{std::numeric_limits<ValueID::base_type>::max()};
+
+// Dictionary is a specific segment type that stores all its values in a vector
+template <typename T>
+class DictionarySegment : public BaseSegment {
+ public:
+  /**
+   * Creates a Dictionary segment from a given value segment.
+   */
+  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment) :
+  _dictionary(std::make_shared<std::vector<T>>()) {
```

Comment:
Yes we use the internal characteristics here. I was relatively sure that the set conversion would be much faster than using the two algorithms, but after reading a little more on it, performance seems to strongly depend on the underlying data (set handles duplicates better, sort+unique is faster without many duplicates).
However, we will not change the code here, but we will add a comment explaining the characteristics of the set, making the code easier to understand

Reply:
What is your intention to not change the code here? As you said, your solution is not faster than using std::unique and std::sort but harder to understand. For me, there is every reason to change it.

Reply:
If I would write this code again knowing that the performance is the same, I'd probably use the other variant. However, since it already happened and the code is only for learning purposes, I do not feel like changing it

Reply:
Ok. We talked about this in the lecture and I'm appeased. Apologies for being so harsh :)